### PR TITLE
use docker auth lib #ferengi

### PR DESCRIPTION
**Pull request by termie**

### CODE REVIEW

File: docker/box.go [hunk 1]
```diff
@@ -538,15 +529,30 @@ func (b *DockerBox) Fetch(ctx context.Context, env *util.Environment) (*docker.I
...
+		options.Repository = env.Interpolate(b.repository)
+	}
+	var authConfig docker.AuthConfiguration
+	if b.config.Auth != nil {
+		authConfig = docker.AuthConfiguration{
+			Username: authenticator.Username(),
+			Password: authenticator.Password(),
+		}
+	}
+	authConfig = docker.AuthConfiguration{
```
Comment (termie) [inline docker/box.go]: this always overrides teh stuff right above it

File: docker/docker.go [hunk 2]
```diff
@@ -927,6 +798,45 @@ func (s *DockerPushStep) InitEnv(env *util.Environment) {
...
+	}
+
+	if awsRegistryID, ok := s.data["aws-registry-id"]; ok {
+		regID := env.Interpolate(awsRegistryID)
+		s.awsRegistryID = regID
+	}
+	var auther auth.Authenticator
+	if s.awsSecretKey != "" {
+		auther = auth.NewAmazonAuth(s.awsRegistryID, s.awsAccessKey, s.awsSecretKey, s.awsRegion, s.awsStrictAuth)
+	} else {
```
Comment (termie) [inline docker/docker.go]: looks like we want to make authenticators here

**Comment by faiq** [comment]
with the addition of the tests from the last 2 commits, i feel like this is ready to be merged

File: wercker.yml [hunk 3]
```diff
@@ -213,3 +213,45 @@ run-tests:
                 unset WERCKER_BUILD_ID
                 cd $CODE_PATH
                 ./test-all.sh
+
+push-to-ecr:
```
Comment (termie) [inline wercker.yml]: this stuff is all extra, now, yeah?

File: docker/docker.go [hunk 4]
```diff
@@ -927,6 +776,48 @@ func (s *DockerPushStep) InitEnv(env *util.Environment) {
...
+	//build auther
+	var opts dockerauth.CheckAccessOptions
+	if username, ok := s.data["username"]; ok {
+		opts.Username = env.Interpolate(username)
+	}
+	if password, ok := s.data["password"]; ok {
+		opts.Password = env.Interpolate(password)
+	}
+	if awsAccessKey, ok := s.data["aws-access-key"]; ok {
+		ak := env.Interpolate(awsAccessKey)
```
Comment (termie) [inline docker/docker.go]: not to be pendantic but let's just assign the var in one line

File: test-all.sh [hunk 5]
```diff
@@ -125,6 +125,11 @@ runTests() {
   basicTest "checkpoint, part 1" build --docker-local --enable-dev-steps "$testsDir/checkpoint" || return 1
   basicTestFail "checkpoint, part 2" build --docker-local --checkpoint foo "$testsDir/checkpoint" || return 1
   basicTest "checkpoint, part 3" build --docker-local --enable-dev-steps --checkpoint foo "$testsDir/checkpoint" || return 1
+
+  # fetching and pushing
+  basicTest "fetch from amazon" build "$testsDir/amzn-test" || return 1
```
Comment (termie) [inline test-all.sh]: can you put some checks around these to skip if the appropriate env vars aren't around?

**Comment by termie** [comment]
boom :lgtm:

**Pull request merged**